Improve namespace packaging guide

chrysle opened the pull request:
This addresses the linked issues, and adds a `pyproject.toml` example for how to find native or implicit namespace packages with `setuptools`.

- Closes #1222
- Closes #689
- Closes #1305

<!-- readthedocs-preview python-packaging-user-guide start -->
----
:books: Documentation preview :books:: https://python-packaging-user-guide--1379.org.readthedocs.build/en/1379/

<!-- readthedocs-preview python-packaging-user-guide end -->

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -86,13 +77,14 @@ structure:
-    setup.py
-    mynamespace/
+    mynamespace-subpackage-a/
+        setup.py # AND/OR pyproject.toml, setup.cfg
```
I would for sure not put `setup.py` in first place here. My personal preference is to only put standardized things in `pyproject.toml` so I would put `setup.cfg` in first place, but I guess the preferred way in the community is to put as much as possible in `pyproject.toml` so that is probably what should be done here as well.

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -112,8 +104,31 @@ list all packages in your :file:`setup.py`. For example:
     )
 
+The same can be accomplished by replacing the :file:`setup.py` in the
+namespace packages' parent directory with a :file:`pyproject.toml`,
```
```suggestion
project's source tree root directory with a :file:`pyproject.toml`,
```

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -125,8 +140,25 @@ the `native namespace package example project`_.
+Legacy namespace packages
+-------------------------
+
+These to methods, that were used to create namespace packages prior to :pep:`420`,
```
```suggestion
These two methods, that were used to create namespace packages prior to :pep:`420`,
```

sinoroc reviewed (commented):
Maybe it is a bigger change... but have you considered migrating the examples in this document to `src` layout?

As far as I know, with the `src` layout the build backends are able to automatically find the packages (namespace and regular), no configuration needed.

chrysle commented:
That #@**-%$!!! pre-commit blames me for typos my PR doesn't include!

sinoroc commented:
> That #@**-%$!!! pre-commit blames me for typos my PR doesn't include!

That is surprising. I have no idea why _codespell_ did not trigger on those issues before...

chrysle commented:
> I have no idea why _codespell_ did not trigger on those issues before...

Well it did, but only **after** #1372 was merged. I guess that useless pre-commit-ci wasn't configured at the time that PR was opened.

chrysle commented:
Thanks for reviewing, I addressed your comments! 

> Maybe it is a bigger change... but have you considered migrating the examples in this document to `src` layout?

I was unsure if I shouldn't just leave the examples under the _Legacy namespace packages_ section alone, considering they're "archived"?

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -57,52 +59,73 @@ import object short).
-            module.py
+    mynamespace-subpackage-a/
+        pyproject.toml # AND/OR setup.py, setup.cfg
+        src/
+            mynamespace/ # namespace package
+            # No __init__.py here.
+                subpackage_a/
+                    # Sub-packages have an __init__.py.
```
I would align indentation like this:

```suggestion
        src/
            mynamespace/ # namespace package
                # No __init__.py here.
                subpackage_a/
                    # Sub-packages have an __init__.py.
```

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -57,52 +59,73 @@ import object short).
-:func:`setuptools.find_packages` won't find the sub-package. You must
-use :func:`setuptools.find_namespace_packages` instead or explicitly
-list all packages in your :file:`setup.py`. For example:
+The ``src-layout`` directory structure allows automatic discovery of packages
+by most :term:`build backends <Build Backend>`. If however you want to manage
```
Add link to "[_src layout vs flat layout_](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/)"? It does not seem to have a target yet though, so it might need to be added.

chrysle replied:
Done!

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -57,52 +59,73 @@ import object short).
+
+.. code-block:: ini
+
+   [options]
```
```suggestion
    [options]
```

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -111,9 +134,13 @@ list all packages in your :file:`setup.py`. For example:
         ...
-        packages=find_namespace_packages(include=['mynamespace.*'])
+        packages=find_namespace_packages(where='src/', include=['mynamespace.subpackage_a']),
+        package_dir={"": "src"},
```
Fixing quotes for consistency of this clock of code.
Maybe we should have some formatting rules for the whole guide?

```suggestion
        package_dir={'': 'src'},
```

chrysle replied:
Yes, we might dig out a hook somewhere.

sinoroc replied:
Something that formats Python code embedded inside reStructuredText? No idea if that exists, but maybe it does exist, I have no idea.

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -57,52 +59,73 @@ import object short).
+            mynamespace/ # namespace package
+                # No __init__.py here.
+                subpackage_a/
+                    # Sub-packages have an __init__.py.
```
```suggestion
                    # Regular import packages have an __init__.py.
```

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -57,52 +59,73 @@ import object short).
+
+.. code-block:: toml
+
+    [build-system]
+    requires = ["setuptools", "setuptools-scm"]
+    build-backend = "setuptools.build_meta"
+
```
If the goal of this code block is to show how to configure setuptools to handle namespace packages, then maybe we can skip the `build-system` table. What do you think?

chrysle replied:
How about the following?

```suggestion
    [build-system]
    ...

```

sinoroc replied:
I tend do write `# ...` or `# [...]` in such cases. But I am fine with anything, even the whole build system table, I don't mind much, it is just nit-picking.

chrysle replied:
Using my suggestion since that comes to you use in several other places.

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -138,22 +182,24 @@ To create a pkgutil-style namespace package, you need to provide an
+            mynamespace/
+                __init__.py  # Namespace package __init__.py
+                subpackage_a/
+                    __init__.py  # Sub-package __init__.py
```
```suggestion
                    __init__.py  # Regular package __init__.py
```

chrysle replied:
I find this a bit confusing, since the namespace package _is_ a sub-package and the root directory is named accordingly?

Probably 
```suggestion
                    __init__.py  # Regular sub-package __init__.py
```

is more like it ;-)

sinoroc replied:
My reasoning for this is that from the point of view of the Python interpreter there are two kinds of importable packages: the regular ones without qualifier and the "namespace" packages.

Whether they are sub-packages or not does not matter.

For example:

```none
.
├── namespace
│   └── regular
│       └── __init__.py
└── regular
    ├── __init__.py
    └── namespace
```

```pycon
>>> import namespace
>>> namespace
<module 'namespace' (<_frozen_importlib_external.NamespaceLoader object at 0x7fb3d531f090>)>
>>> import namespace.regular
>>> namespace.regular
<module 'namespace.regular' from '/tmp/tmp.wRR0QJjvej/namespace/regular/__init__.py'>
>>> import regular
>>> regular
<module 'regular' from '/tmp/tmp.wRR0QJjvej/regular/__init__.py'>
>>> import regular.namespace
>>> regular.namespace
<module 'regular.namespace' (<_frozen_importlib_external.NamespaceLoader object at 0x7fb3d4b6af90>)>
```

I know that this is valid for PEP420 native namespace packages, and this example is not PEP420, but from my point of view we should try to stick to this terminology.

I guess I am fine with "regular sub-package".

chrysle replied:
Let's do it like you suggest, then. I think I'm the one who's nit-picking.

sinoroc commented on source/guides/packaging-namespace-packages.rst:
```diff
@@ -183,22 +229,24 @@ To create a pkg_resources-style namespace package, you need to provide an
+            mynamespace/
+                __init__.py  # Namespace package __init__.py
+                subpackage_a/
+                    __init__.py  # Sub-package __init__.py
```
```suggestion
                    __init__.py  # Regular package __init__.py
```

sinoroc reviewed (commented):
Some more suggestions. Sorry that I am bringing yet another round of those instead of all at once... :/

sinoroc reviewed (approved):
I approve, no reason to block this.

chrysle commented:
@sinoroc Thank you for you thorough review!

willingc reviewed (approved):
This looks good to me. Thanks!
